subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2213 @@
+	case -ENODEV:
+	case -ENOTSUP:
+	case -ENOSYS:
+		data->err = ERR_INVALID_REQUEST;
+		break;
+	default:
+		data->err = ERR_UNKNOWN;
+		break;
+	}
+
+	return ret;
+}
+
+static int uvc_control_errno(uint8_t request, struct net_buf *buf, int err)
+{
+	switch (request) {
+	case GET_INFO:
+		return uvc_buf_add_int(buf, 1, INFO_SUPPORTS_GET);
+	case GET_CUR:
+		return uvc_buf_add_int(buf, 1, err);
+	default:
+		LOG_WRN("Unsupported request type %u", request);
+		return -ENOTSUP;
+	}
+}
+
+static int uvc_control(const struct device *dev, const struct usb_setup_packet *const setup,
+		       struct net_buf *buf)
+{
+	struct uvc_data *data = dev->data;
+	uint8_t ifnum = (setup->wIndex >> 0) & 0xff;
+	uint8_t unit_id = (setup->wIndex >> 8) & 0xff;
+	uint8_t request = setup->bRequest;
+
+	LOG_DBG("Host sent a %s request",
+		request == SET_CUR ? "SET_CUR" : request == GET_CUR ? "GET_CUR" :
+		request == GET_MIN ? "GET_MIN" : request == GET_MAX ? "GET_MAX" :
+		request == GET_RES ? "GET_RES" : request == GET_LEN ? "GET_LEN" :
+		request == GET_DEF ? "GET_DEF" : request == GET_INFO ? "GET_INFO" : "bad");
+
+	if (setup->wLength > buf->size) {
+		LOG_ERR("wLength %u larger than %u bytes", setup->wLength, buf->size);
+		return -ENOMEM;
+	}
+	buf->size = setup->wLength;
+
+	/* VideoStreaming requests */
+
+	for (struct uvc_stream *strm = data->streams; strm->dev != NULL; strm++) {
+		if (ifnum == strm->desc->if_vs.bInterfaceNumber) {
+			return uvc_control_vs(strm, setup, buf);
+		}
+	}
+
+	/* VideoControl requests */
+
+	if (ifnum != data->desc->if_vc.bInterfaceNumber) {
+		LOG_WRN("Interface %u not found", ifnum);
+		data->err = ERR_INVALID_UNIT;
+		return -ENOTSUP;
+	}
+
+	if (unit_id == 0) {
+		return uvc_control_errno(request, buf, data->err);
+	}
+
+	for (struct uvc_stream *strm = data->streams; strm->dev != NULL; strm++) {
+		if (unit_id == strm->desc->if_vc_ct.bTerminalID) {
+			return uvc_control_call(dev, strm, setup, buf, uvc_control_map_ct);
+		} else if (unit_id == strm->desc->if_vc_su.bUnitID) {
+			return uvc_control_call(dev, strm, setup, buf, uvc_control_map_su);
+		} else if (unit_id == strm->desc->if_vc_pu.bUnitID) {
+			return uvc_control_call(dev, strm, setup, buf, uvc_control_map_pu);
+		} else if (unit_id == strm->desc->if_vc_xu.bUnitID) {
+			return uvc_control_call(dev, strm, setup, buf, uvc_control_map_xu);
+		}
+	}
+
+	LOG_WRN("Could not handle command for bUnitID %u", unit_id);
+	data->err = ERR_INVALID_UNIT;
+	return -ENOTSUP;
+}
+
+static int uvc_control_to_host(struct usbd_class_data *const c_data,
+			       const struct usb_setup_packet *const setup,
+			       struct net_buf *const buf)
+{
+	errno = uvc_control(usbd_class_get_private(c_data), setup, buf);
+	return 0;
+}
+
+static int uvc_control_to_dev(struct usbd_class_data *const c_data,
+			      const struct usb_setup_packet *const setup,
+			      const struct net_buf *const buf)
+{
+	errno = uvc_control(usbd_class_get_private(c_data), setup, (struct net_buf *)buf);
```

Comment:
Sorry, I did not see this earlier. You must not convert the type of the net_buf pointer here. The buffer for the to-device request should be read-only, and the handlers must have constant-data pointer parameter consistently in all subsequent routines. The to-device and to-host requests must be handled separately.

Reply:
No problem for my side!
I hope the new approach is more suitable. If not, I am fine with changing it until it is.